text/3389-manifest-lint.md

```diff
@@ -0,0 +1,398 @@
+```toml
+[lints]
+workspace = true
+
+[workspace.lints]
+unsafe = "forbid"
+```
+
+## Documentation Updates
+
+## The `lints` section
+
+*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*
+
+Override the default level of lints from different tools by assigning them to a new level in a
+table, for example:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+
+This is short-hand for:
+```toml
+[lints.rust]
+unsafe = { level = "forbid", priority = 0 }
+```
+
+`level` corresponds to the lint levels in `rustc`:
+- `forbid`
+- `deny`
+- `warn`
+- `allow`
+
+`priority` is a signed integer that controls which lints or lint groups override other lint groups:
+- lower (particularly negative) numbers have lower priority, being overridden
+  by higher numbers, and show up first on the command-line to tools like
+  `rustc`
+
+To know which table under `[lints]` a particular lint belongs under, it is the part before `::` in the lint
+name.  If there isn't a `::`, then the tool is `rust`.  For example a warning
+about `unsafe` would be `lints.rust.unsafe` but a lint about
+`clippy::enum_glob_use` would be `lints.clippy.enum_glob_use`.
+
+## The `lints` table
+
+*as a new [`[workspace]` entry](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section)*
+
+The `workspace.lints` table is where you define lint configuration to be inherited by members of a workspace.
+
+Specifying a workspace lint configuration is similar to package lints.
+
+Example:
+
+```toml
+# [PROJECT_DIR]/Cargo.toml
+[workspace]
+members = ["crates/*"]
+
+[workspace.lints.rust]
+unsafe = "forbid"
+```
+
+```toml
+# [PROJECT_DIR]/crates/bar/Cargo.toml
+[package]
+name = "bar"
+version = "0.1.0"
+
+[lints]
+workspace = true
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When parsing a manifest, cargo will resolve workspace inheritance for
+`lints.workspace = true` as it does with basic fields, when `workspace` is
+present, no other fields are allowed to be present.  This precludes having the
+package override the workspace on a lint-by-lint basis.
+
+cargo will contain a mapping of tool to underlying command (e.g. `rust` to
+`rustc`, `clippy` to `rustc` when clippy is the driver, `rustdoc` to
+`rustdoc`).  When running the underlying command for the specified package,
+cargo will transform the lints from `lint = level` to `--level lint`, sort them
+by priority and then lint name, and pass them on the command line before other
+configuration, `RUSTFLAGS`, allowing user configuration to override package
+configuration.  These flags will be fingerprinted so changing them will cause a
+rebuild only for the commands where they are used.
+
+Note that this means that `[lints]` does not affect dependencies which is
+normally not an issue due to `--cap-lints` being used for dependencies.
+
+Initially, the only supported tools will be:
+- `rust`
+- `clippy`
+- `rustdoc`
```

Comment:
I'm not sure I understand what is meant by being ignored.

If an unknown tool is passed to rustc, it gives a hard error ([E0602](https://doc.rust-lang.org/error_codes/E0602.html)). 

Rustc only ignores lint paths for tools it knows about (clippy, rustc, rustfmt, and rustdoc) or are registered (a bit complicated). 

As for my comment about namespaces, I was mostly concerned about the idea that cargo is converting `::` to `.`. Does it convert all `::`? If just the first `::`, might that cause confusion? (This is hypothetical, since AFAIK none of the current tools bother with additional `::` separators.)

As for restricting the set of tools names, I think I would also lean towards "forward it to rustc and rustdoc and let them figure it out".  Though that may further cause problems with the 1-letter difference between `rustc` and `rust`, where if a user mistakenly adds a `rustc` table, it will essentially be ignored.

Reply:
> I'm not sure I understand what is meant by being ignored.

`rustc` only splits on the first `::` and treats the rest of the "Path" of the lint as a plain string. So it doesn't parse the whole lint name into a `Path` and treats it like module namespaces. 



> As for my comment about namespaces, I was mostly concerned about the idea that cargo is converting `::` to `.`.

Cargo doesn't convert `::` to `.`, but rather `.` to `::`. But the `.` it converts is part of the TOML syntax and not part of a string. If you define a lint name as a string `"mytool::foo::bar"`  cargo shouldn't touch it.